imgui/src/input_widget.rs

```diff
@@ -199,6 +199,297 @@ extern "C" fn resize_callback(data: *mut sys::ImGuiInputTextCallbackData) -> c_i
     }
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+pub enum CallbackType {
+    Completion,
+    History,
+    CharacterFilter,
+    Edit,
+    Resize,
+}
+// tighten the data provided to each event type
+// so that the user can understand what they need to use to do what they want
+// don't allow resize to be registered, and just let the normal callback
+
+pub struct CallbackData {
+    inner: *mut sys::ImGuiInputTextCallbackData,
+}
+
+impl CallbackData {
+    // todo make this more intuitive
+    pub fn insert_characters(&mut self, pos: i32, text: &ImStr) {
+        unsafe {
+            sys::ImGuiInputTextCallbackData_InsertChars(
+                self.inner,
+                pos,
+                text.as_ptr(),
+                ptr::null(), // imgui will read strlen for us based on nul character, though if we want to calc this our selves we can
+            );
+        }
+    }
+    // TODO make this more intuitive
+    // things like deleting to the last X should be easy todo?
+    pub fn delete_characters(&mut self, pos: i32, bytes_count: i32) {
+        unsafe { sys::ImGuiInputTextCallbackData_DeleteChars(self.inner, pos, bytes_count) }
+    }
+
+    pub fn has_selection(&mut self) -> bool {
+        unsafe { sys::ImGuiInputTextCallbackData_HasSelection(self.inner) }
+    }
+
+    pub fn clear_selection(&mut self) {
+        unsafe { sys::ImGuiInputTextCallbackData_ClearSelection(self.inner) }
+    }
+
+    pub fn select_all(&mut self) {
+        unsafe { sys::ImGuiInputTextCallbackData_SelectAll(self.inner) }
+    }
+
+    pub fn cursor_pos(&self) -> usize {
+        let byte_position = unsafe { (*self.inner).CursorPos } as usize;
+
+        self.byte_to_index(byte_position).unwrap()
+    }
+
+    pub fn set_cursor_pos(&mut self, index: usize) -> bool {
+        if let Some(position) = self.index_to_byte(index) {
+            unsafe {
+                (*self.inner).CursorPos = position as i32;
+            }
+            true
+        } else {
+            false
+        }
+    }
+
+    pub fn event_char(&self) -> char {
+        // pretty sure EventChar is utf16 but i need to double check
+        char::decode_utf16(std::iter::once(unsafe { (*self.inner).EventChar }))
```

Comment:
std::char::decode_utf16 works here without the unstable feature.

Reply:
Fixed